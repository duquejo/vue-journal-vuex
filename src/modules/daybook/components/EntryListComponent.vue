<template>
  <div class="entry-list-container">
    <div class="px-2 pt-2">
      <input v-model="term" type="text" class="form-control" placeholder="Search entry" />
    </div>

    <div class="mt-2 d-flex flex-column">
      <button @click="$router.push({ name: 'entry', params: { id: 'new' } })" 
        class="btn btn-primary mx-2">
        <i class="fa fa-plus-circle"/>
        New entry
      </button>
    </div>

    <div class="entry-scrollable-area">
      <EntryComponent
        v-for="entry in entriesByTerm"
        :key="entry.id"
        :entry="entry"
      />
    </div>
  </div>
</template>

<script>
import { defineAsyncComponent } from 'vue';
import { mapGetters } from 'vuex';

export default {
  components: {
    EntryComponent: defineAsyncComponent(() => import('./EntryComponent')),
  },
  data() {
    return {
      term: '',
    }
  },
  computed: {
    ...mapGetters('journal', ['getEntriesByTerm']),
    entriesByTerm() {
      return this.getEntriesByTerm( this.term );
    },
  },
}
</script>

<style lang="scss" scoped>
.entry-list-container {
  border-right: 1px solid #c7c8c9;
  height: calc(100vh - 56px);
}
.entry-scrollable-area {
  height: calc(100vh - 110px);
  overflow: scroll;
}
</style>