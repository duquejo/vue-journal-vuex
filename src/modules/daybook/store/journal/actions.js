import journalApi from "../../../../api/journalApi";

export const loadEntries = async ({ commit }) => {
  const { data } = await journalApi.get('/entries.json');

  if( ! data ) {
    commit('setEntries', []);
    return;
  }
  const entries = [];
  for (let id of Object.keys(data)) {
    entries.push({
      id,
      ...data[id],
    });
  }
  commit('setEntries', entries);
};
export const updateEntry = async ({ commit }, entry ) => {
  const { id, date, picture, text  } = entry;

  const dataToSave = { date, picture, text };

  await journalApi.put(`/entries/${id}.json`, dataToSave );

  dataToSave.id = id;

  commit('updateEntry', { ...dataToSave });
};
export const createEntry = async ({ commit }, entry ) => {
  const { date, picture, text } = entry;

  const dataToSave = { date, picture, text };

  const { data } = await journalApi.post('/entries.json', dataToSave );
  commit('addEntry', { id: data.name, ...dataToSave });
  return data.name;
};
export const deleteEntry = async ({ commit }, id ) => {
  await journalApi.delete(`/entries/${id}.json` );
  commit('deleteEntry', id );
  return id;
};