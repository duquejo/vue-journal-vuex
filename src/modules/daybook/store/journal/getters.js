export const getEntriesByTerm = (/*state*/) => {
};
export const getEntryById = (/*state*/) => {
};