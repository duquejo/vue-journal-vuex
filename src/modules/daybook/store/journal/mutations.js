export const setEntries = (/*state*/) => {
};
export const setLoading = (/*state*/) => {
};
export const updateEntry = (/*state*/) => {
};
export const addEntry = (/*state*/) => {
};