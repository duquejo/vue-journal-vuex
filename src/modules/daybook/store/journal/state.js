export default () => ({
  isLoading: true,
  entries: [
    {
      id: new Date().getTime(), // 12315646734
      date: new Date().toDateString(), // Sat 23, July
      text: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit. Soluta recusandae iusto harum sint maxime laborum.',
      picture: null,
    },
    {
      id: new Date().getTime() + 1000, // 12315646734
      date: new Date().toDateString(), // Sat 23, July
      text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Molestias, blanditiis.',
      picture: null,
    },
    {
      id: new Date().getTime() + 2000, // 12315646734
      date: new Date().toDateString(), // Sat 23, July
      text: 'Lorem ipsum dolor sit amet.',
      picture: null,
    },
  ]
});