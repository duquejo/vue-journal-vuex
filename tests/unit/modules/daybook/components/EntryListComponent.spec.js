import { createStore } from 'vuex';
import journal from '@/modules/daybook/store/journal';
import { shallowMount } from '@vue/test-utils';
import EntryListComponent from '@/modules/daybook/components/EntryListComponent';
import { journalState } from '../../../mocks/test-journal-state';

describe('EntryListComponent.vue', () => {

  const routerMock = {
    push: jest.fn(),
  };

  const createVuexStore = (initialState) => createStore({
    modules: {
      journal: {
        ...journal,
        state: { ...initialState },
      },
    },
  });

  const store = createVuexStore(journalState);

  let wrapper;

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryListComponent,{
      global: {
        mocks: {
          $router: routerMock,
        },
        plugins: [ store ],
      },
    });
  });

  it('should call getEntriesByTerm without term and it should return 2 entries.', () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.findAll('entry-component-stub')).toHaveLength(2);
  });

  it('should call getEntriesByTerm and it should filter the provided entries.', async () => {
    const input = wrapper.find('input');
    await input.setValue('development');
    expect(wrapper.findAll('entry-component-stub')).toHaveLength(1);
  });

  it('should the new button redirect to the following /new subview.', () => {
    wrapper.find('button').trigger('click');
    expect(routerMock.push).toHaveBeenCalled();
    expect(routerMock.push).toHaveBeenCalledTimes(1);
    expect(routerMock.push).toHaveBeenCalledWith({ name: 'entry', params: { id: 'new' }});
  });
});