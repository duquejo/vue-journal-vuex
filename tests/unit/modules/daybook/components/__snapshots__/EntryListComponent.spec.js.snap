// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryListComponent.vue should call getEntriesByTerm without term and it should return 2 entries. 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Search entry"></div>
  <div class="mt-2 d-flex flex-column"><button class="btn btn-primary mx-2"><i class="fa fa-plus-circle"></i> New entry </button></div>
  <div class="entry-scrollable-area">
    <entry-component-stub entry="[object Object]"></entry-component-stub>
    <entry-component-stub entry="[object Object]"></entry-component-stub>
  </div>
</div>
`;
